network_wrangler/utils/transit.py

```diff
@@ -0,0 +1,3087 @@
+            )
+
+    # verify model_node_id, f'match_distance_{crs_units}' and 'valid_match' are not in feed_tables['stops']
+    assert "model_node_id" not in feed_tables["stops"].columns
+    assert f"match_distance_{crs_units}" not in feed_tables["stops"].columns
+    assert "valid_match" not in feed_tables["stops"].columns
+
+    # Update feed_tables['stops'] by merging the updates
+    feed_tables["stops"] = feed_tables["stops"].merge(
+        bus_stops_gdf[
+            [
+                "stop_id",
+                "model_node_id",
+                f"match_distance_{crs_units}",
+                "stop_lon",
+                "stop_lat",
+                "geometry",
+                "valid_match",
+            ]
+        ],
+        on="stop_id",
+        how="left",
+        suffixes=("", "_bus"),
+        validate="one_to_one",
+    )
+    # Only update stop location for valid match
+    feed_tables["stops"].loc[feed_tables["stops"]["valid_match"] == True, "stop_lon"] = (
+        feed_tables["stops"]["stop_lon_bus"]
+    )
+    feed_tables["stops"].loc[feed_tables["stops"]["valid_match"] == True, "stop_lat"] = (
+        feed_tables["stops"]["stop_lat_bus"]
+    )
+    feed_tables["stops"].loc[feed_tables["stops"]["valid_match"] == True, "geometry"] = (
+        feed_tables["stops"]["geometry_bus"]
+    )
+    # Drop bus-specific columns
+    feed_tables["stops"].drop(
+        columns=["stop_lon_bus", "stop_lat_bus", "geometry_bus"], inplace=True
+    )
+
+    WranglerLogger.debug(
+        f"After merging with bus_stops_gdf, feed_tables['stops']:\n{feed_tables['stops']}"
+    )
+    if trace_shape_ids:
+        for trace_shape_id in trace_shape_ids:
+            WranglerLogger.debug(
+                f"trace feed_tables['stops'] for {trace_shape_id}:\n"
+                f"{feed_tables['stops'].loc[feed_tables['stops']['shape_ids'].apply(lambda x, tid=trace_shape_id: isinstance(x, list) and tid in x)]}"
+            )
+
+    # summary of match distances
+    WranglerLogger.debug(
+        f"\n{feed_tables['stops'][['valid_match', f'match_distance_{crs_units}']].describe()}"
+    )
+
+    # Update shapes table similarly
+    WranglerLogger.debug(f"feed_tables['shapes']:\n{feed_tables['shapes']}")
+    # columns: shape_id, shape_pt_lat, shape_pt_lon, shape_pt_sequence, shape_dist_traveled, geometry
+    #   trip_id, direction_id, route_id, agency_id, route_short_name, route_type, agency_name,
+    #   match_distance_feet, stop_id, stop_name, stop_sequence
+    # Note: this is adding: 'model_node_id'
+    feed_tables["shapes"] = pd.merge(
+        left=feed_tables["shapes"],
+        right=bus_stops_gdf[
+            [
+                "stop_id",
+                "model_node_id",
+                f"match_distance_{crs_units}",
+                "stop_lon",
+                "stop_lat",
+                "geometry",
+                "valid_match",
+            ]
+        ],
+        on="stop_id",
```

Comment:
I'm not sure I get how shapes table ends up with a good `stop_id`? I might need to reread and figure out...